File: app/api/v1/portal.py
"""
Client Portal API endpoints.
Provides secure access for clients to respond to requests.
"""

from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File
from sqlalchemy.orm import Session
from datetime import datetime

from app.api.deps import get_db
from app.db.postgres.repositories import PostgresRequestRepository
from app.services.request_service import RequestService
from app.schemas.portal import (
    PortalValidateResponse,
    PortalRequestResponse,
    PortalRequestInfo,
    PortalRespondRequest,
    PortalRespondResponse,
    PortalUploadResponse,
)
from app.models.request import RequestStatus, REQUEST_TYPE_CONFIG, RequestType

router = APIRouter()


# Human-readable request type names
REQUEST_TYPE_DISPLAY = {
    "insurance_coverage_claim": "Insurance Coverage Claim",
    "wrong_insurance_refile": "Wrong Insurance / Refile",
    "auto_accident_3rd_party": "Auto Accident / 3rd Party",
    "medicaid_medicare_question": "Medicaid/Medicare Question",
    "paid_direct_to_provider": "Paid Direct to Provider",
    "on_payment_plan": "On Payment Plan",
    "payment_posted_wrong": "Payment Posted Wrong",
    "not_our_patient": "Not Our Patient / No Service",
    "identity_theft_fraud": "Identity Theft / Fraud",
    "itemized_bill_request": "Itemized Bill Request",
    "validation_package_request": "Validation Package Request",
    "statement_resend": "Statement Resend",
    "overcharged_balance_incorrect": "Overcharged / Balance Incorrect",
    "service_cancelled_not_billed": "Service Cancelled / Not Billed",
}


@router.get("/validate/{token}", response_model=PortalValidateResponse)
async def validate_portal_token(
    token: str,
    db: Session = Depends(get_db)
):
    """
    Validate a portal access token.
    Returns whether the token is valid and not expired.
    """
    repo = PostgresRequestRepository(db)
    request = await repo.get_by_portal_token(token)

    if not request:
        return PortalValidateResponse(
            is_valid=False,
            error="Invalid or expired token"
        )

    # Check expiration
    if request.portal_token_expires_at and request.portal_token_expires_at < datetime.utcnow():
        return PortalValidateResponse(
            is_valid=False,
            error="Token has expired"
        )

    # Check if request is in correct status
    if request.status not in [RequestStatus.SENT.value, RequestStatus.RESPONDED.value]:
        return PortalValidateResponse(
            is_valid=False,
            error="Request is not awaiting response"
        )

    return PortalValidateResponse(
        is_valid=True,
        request_id=request.request_id,
        expires_at=request.portal_token_expires_at
    )


@router.get("/{token}/request", response_model=PortalRequestResponse)
async def get_portal_request(
    token: str,
    db: Session = Depends(get_db)
):
    """
    Get request information for the portal.
    Returns limited, PHI-safe information for the client to respond.
    """
    repo = PostgresRequestRepository(db)
    request = await repo.get_by_portal_token(token)

    if not request:
        return PortalRequestResponse(
            is_valid=False,
            error="Invalid or expired token"
        )

    # Check expiration
    if request.portal_token_expires_at and request.portal_token_expires_at < datetime.utcnow():
        return PortalRequestResponse(
            is_valid=False,
            error="Token has expired"
        )

    # Build PHI-safe request info
    request_type_value = request.request_type if request.request_type else "unknown"
    request_info = PortalRequestInfo(
        request_id=request.request_id,
        request_type=request_type_value,
        request_type_display=REQUEST_TYPE_DISPLAY.get(request_type_value, request_type_value),
        account_reference=request.account_reference,
        created_at=request.created_at,
        status=request.status,
    )

    return PortalRequestResponse(
        is_valid=True,
        request=request_info,
        client_name=request.client.name if request.client else None
    )


@router.post("/{token}/respond", response_model=PortalRespondResponse)
async def submit_portal_response(
    token: str,
    response_data: PortalRespondRequest,
    db: Session = Depends(get_db)
):
    """
    Submit a response via the portal.
    Client provides resolution type, notes, and any additional info.
    """
    repo = PostgresRequestRepository(db)
    request = await repo.get_by_portal_token(token)

    if not request:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Invalid or expired token"
        )

    # Check expiration
    if request.portal_token_expires_at and request.portal_token_expires_at < datetime.utcnow():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Token has expired"
        )

    # Check status
    if request.status != RequestStatus.SENT.value:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Request has already been responded to or is not awaiting response"
        )

    # Record the response
    service = RequestService(db)
    await service.record_client_response(
        request.request_id,
        response_data.model_dump()
    )

    return PortalRespondResponse(
        success=True,
        message="Response submitted successfully. The agency will review and follow up.",
        request_id=request.request_id
    )


@router.post("/{token}/upload", response_model=PortalUploadResponse)
async def upload_portal_file(
    token: str,
    file: UploadFile = File(...),
    db: Session = Depends(get_db)
):
    """
    Upload a file via the portal.
    Accepts documents like EOBs, ledgers, proof of payment, etc.
    """
    repo = PostgresRequestRepository(db)
    request = await repo.get_by_portal_token(token)

    if not request:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Invalid or expired token"
        )

    # Check expiration
    if request.portal_token_expires_at and request.portal_token_expires_at < datetime.utcnow():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Token has expired"
        )

    # Validate file type
    allowed_types = [
        "application/pdf",
        "image/jpeg",
        "image/png",
        "image/gif",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    ]

    if file.content_type not in allowed_types:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"File type not allowed. Allowed types: PDF, JPEG, PNG, GIF, DOC, DOCX"
        )

    # Check file size (max 10MB)
    max_size = 10 * 1024 * 1024
    contents = await file.read()
    if len(contents) > max_size:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="File too large. Maximum size is 10MB."
        )

    # TODO: Save file to storage (local, S3, or Supabase storage)
    # For now, return a placeholder response

    return PortalUploadResponse(
        success=True,
        filename=file.filename,
        file_id=0,  # TODO: Return actual attachment ID
        message="File uploaded successfully"
    )

File: src/api/portal.ts
/**
 * Portal API Service
 * Handles client portal operations (public-facing, token-based auth)
 */

import type { RequestType, ResolutionCode } from './requests';

const API_BASE_URL = import.meta.env.VITE_API_URL || '/api/v1';

export interface PortalRequestInfo {
  request_id: string;
  account_reference: string;
  request_type: RequestType;
  request_type_display: string;
  file_id?: string;
  balance?: string;
  details?: Record<string, unknown>;
  created_at: string;
  expires_at?: string;
  has_responded?: boolean;
  status?: string;
  client_name?: string;
}

// Backend response wrapper
interface PortalRequestResponse {
  is_valid: boolean;
  request?: {
    request_id: string;
    request_type: string;
    request_type_display: string;
    account_reference: string;
    file_id?: string;
    balance?: string;
    inquiry_details?: string;
    created_at: string;
    status: string;
  };
  client_name?: string;
  expires_at?: string;
  error?: string;
}

export interface PortalResponseData {
  resolution: ResolutionCode;
  notes?: string;
  additional_info?: Record<string, unknown>;
}

export interface PortalUploadResponse {
  id: number;
  filename: string;
  size: number;
  uploaded_at: string;
}

export interface PortalSubmitResponse {
  success: boolean;
  message: string;
  request_id: string;
}

class PortalAPI {
  private baseUrl: string;

  constructor() {
    this.baseUrl = API_BASE_URL;
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...options.headers,
    };

    const response = await fetch(`${this.baseUrl}${endpoint}`, {
      ...options,
      headers,
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || error.message || `Error: ${response.status}`);
    }

    if (response.status === 204) {
      return {} as T;
    }

    return response.json();
  }

  /**
   * Get request information for portal (no auth required, uses token)
   */
  async getPortalRequest(token: string): Promise<PortalRequestInfo> {
    const response = await this.request<PortalRequestResponse>(`/portal/${token}/request`);

    if (!response.is_valid || !response.request) {
      throw new Error(response.error || 'Invalid or expired token');
    }

    // Map backend response to frontend interface
    return {
      request_id: response.request.request_id,
      account_reference: response.request.account_reference,
      request_type: response.request.request_type as RequestType,
      request_type_display: response.request.request_type_display,
      file_id: response.request.file_id,
      balance: response.request.balance,
      details: response.request.inquiry_details ? { notes: response.request.inquiry_details } : undefined,
      created_at: response.request.created_at,
      status: response.request.status,
      client_name: response.client_name,
      expires_at: response.expires_at,
      has_responded: response.request.status === 'RESPONDED',
    };
  }

  /**
   * Submit client response via portal
   */
  async submitResponse(token: string, data: PortalResponseData): Promise<PortalSubmitResponse> {
    return this.request<PortalSubmitResponse>(`/portal/${token}/respond`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  /**
   * Upload document via portal
   */
  async uploadDocument(token: string, file: File): Promise<PortalUploadResponse> {
    const formData = new FormData();
    formData.append('file', file);

    const response = await fetch(`${this.baseUrl}/portal/${token}/upload`, {
      method: 'POST',
      body: formData,
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || error.message || 'Upload failed');
    }

    return response.json();
  }

  /**
   * Validate portal token
   */
  async validateToken(token: string): Promise<{ valid: boolean; expires_at?: string }> {
    return this.request<{ valid: boolean; expires_at?: string }>(`/portal/validate/${token}`);
  }

  /**
   * Get list of uploaded documents for this portal session
   */
  async getUploadedDocuments(token: string): Promise<PortalUploadResponse[]> {
    return this.request<PortalUploadResponse[]>(`/portal/${token}/documents`);
  }

  /**
   * Delete an uploaded document
   */
  async deleteDocument(token: string, documentId: number): Promise<void> {
    return this.request<void>(`/portal/${token}/documents/${documentId}`, {
      method: 'DELETE',
    });
  }
}

export const portalAPI = new PortalAPI();
export default portalAPI;
